ast/src/main/java/com/graphicsfuzz/common/ast/decl/ArrayInfo.java

```diff
@@ -17,37 +17,58 @@
 package com.graphicsfuzz.common.ast.decl;
 
 import com.graphicsfuzz.common.ast.IAstNode;
+import com.graphicsfuzz.common.ast.expr.Expr;
+import com.graphicsfuzz.common.ast.expr.IntConstantExpr;
 import com.graphicsfuzz.common.ast.visitors.IAstVisitor;
+import com.graphicsfuzz.common.ast.visitors.UnsupportedLanguageFeatureException;
 import java.util.Optional;
 
 public class ArrayInfo implements IAstNode {
 
-  private Optional<Integer> size;
+  private Optional<Expr> size;
+  private Optional<Expr> originalSize;
 
-  public ArrayInfo(Integer size) {
+  public ArrayInfo(Expr size) {
     this.size = Optional.of(size);
+    this.originalSize = Optional.of(size);
   }
 
   public ArrayInfo() {
     this.size = Optional.empty();
+    this.originalSize = Optional.empty();
   }
 
   public boolean hasSize() {
     return size.isPresent();
   }
 
-  public Integer getSize() {
+  public Integer getSize() throws UnsupportedLanguageFeatureException {
+    if (size.get() instanceof IntConstantExpr) {
+      return ((IntConstantExpr)size.get()).getNumericValue();
+    }
+    throw new UnsupportedLanguageFeatureException("Not a constant expression");
+  }
+
+  public Expr getSizeExpr() {
     return size.get();
   }
 
+  public void setSizeExpr(Expr expr) {
+    this.size = Optional.of(expr);
+  }
+
+  public Expr getOriginalSizeExpr() {
+    return originalSize.get();
+  }
+
   @Override
   public void accept(IAstVisitor visitor) {
     visitor.visitArrayInfo(this);
   }
 
   @Override
   public ArrayInfo clone() {
```

Comment:
Hmm, I think we probably want to clone the original size expression as well, otherwise cloning will change pretty printing.

Could you make a test class, ArrayInfoTest, and add a test that checks that a shader with a foldable array pretty prints to the same thing after cloning?

Reply:
Removed cloning from there. For some reason tests pass without clone() anywhere now. Not sure why. They did fail at some point.